### ISSUE TITLE
[Docs] Update documentation and examples to use RestApplication

### ISSUE THREAD

**Issue by kjdelisle** [strongloop/loopback-next#846]
## Overview
Rewrite the existing examples to leverage the new `RestApplication` class, which should simplify those examples nicely.

## Acceptance Criteria
- [x] Update all loopback.io examples of single-server usage with `RestApplication`
- [x] Update the `loopback/example-getting-started` application to use `RestApplication`

**Comment by bajtos** [comment]
Unfortunately, the acceptance criteria are not specific enough about what kind of changes need to be done - what does it mean to update an application to use `RestApplication`?

In https://github.com/strongloop/loopback-next/pull/955, I am arguing that it's not enough to extend from `RestApplication` and remove `RestComponent` usage, but that we should revert changes that moved sequence/route initialization from constructor to `app.start` method. 

The setup code was moved to async start function after @kjdelisle implemented support for multiple REST servers. I was never happy about that change to be honest. For example, it means that code inspecting application's OpenAPI spec (think of `lb export-api-def`) cannot see all routes until the application is started. Now that we have `RestApplication`, there is no need to use that suboptimal workaround and we can revert back to the original version.

IMO, this second set of changes is required before we can call this task as done.

**Comment by bajtos** [comment]
I decided to go ahead and help you, see https://github.com/strongloop/loopback-next/pull/968. When that pull request is landed, I think the only remaining task is to update our docs accordingly.

**Comment by shimks** [comment]
While I was looking through the docs to find `RestApplication` configs to move, I noticed that I wasn't able to set up any server-level bindings in the constructor. For example, I'm not able to register routes through `app.route(new Route(...))` since application-level context doesn't seem to inherit from server-level context.

Does this mean potentially we should discuss whether `RestApplication` should merge the application-level and server-level contexts together?

**Comment by bajtos** [comment]
> While I was looking through the docs to find `RestApplication` configs to move, I noticed that I wasn't able to set up any server-level bindings in the constructor. For example, I'm not able to register routes through `app.route(new Route(...))` since an application doesn't seem to inherit server-level binding functions.
>
> Does this mean potentially we should discuss whether `RestApplication` should merge the application-level and server-level contexts together? If not, should we create sugar to allow server-level bindings to be made at application-level (like `this.sequence()` in `RestApplication`)?

Good catch, this looks like a bug in `RestApplication` to me.

I see two possible solutions and don't know which of them is better/correct:

 - Modify `app.route` (et all) to delegate the work to `server.route`, where `server` is the single/main instance of the REST server.

 - Modify the place creating server-level contexts to let the server context inherit from the application context. (Conceptually: `server.context = new Context(app)`.) I am surprised this is not already happening!

**Comment by virkt25** [comment]
I personally prefer the idea of having server context set app context as the parent. Maybe there was a reason that wasn't done?

**Comment by kjdelisle** [comment]
@bajtos 
>Good catch, this looks like a bug in RestApplication to me.

RestApplication wasn't meant to be a "merged" context that acts as both application and singular rest server. Its primary purpose is to limit the need for users to import things like `RestComponent` and `RestServer` when making (what we're assuming) will be the most common form of LoopBack 4 app.

If you look at the [constructor](https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/rest-application.ts#L41), you'll see that we provide the `RestComponent` constructor if no components are given.

By extension, we provide a safeguarded version of the `server` function so that people can bind an alternate server if they were to override the components collection passed into the constructor.
This is because replacing the `RestComponent` means you might not automatically receive a `RestServer` instance (depending on what's been provided to replace it).

If retrieval of the server instance is required for configuration, it still demands use of an override of the `start` function in order to retrieve the `RestServer` instance and modify its context accordingly.

@shimks 
With respect to the components changes that have been proposed, you'll probably want to add code that checks the app context for the existence of a `RestComponent` before moving on (since they're now expected to call `app.component` rather than passing it as a constructor parameter).

-------

>Modify the place creating server-level contexts to let the server context inherit from the application context. (Conceptually: server.context = new Context(app).) I am surprised this is not already happening!

This was the original implementation of `RestServer`, but I was asked to componentize it as `RestComponent` and then was asked to make that component instantiate a `RestServer`! 🤦‍♂️ 

If you want to simplify the REST story for everyone in terms of both UX and code complexity, then we should (in a _separate_ issue/PR):
- Disable the instantiation of `RestServer` as a part of the use of `RestComponent`
- Refactor `RestComponent` into a collection of separate providers, that must also be exported
- Remove `RestComponent` **entirely** (as it's no longer necessary with the existence of `RestApplication`)
- Change the `RestApplication` constructor so that it:
  - Instantiates an instance of `RestServer`, passing itself in to define a parent-child link directly
  - Binds that instance using its own `server` function (which should still preserve its existing logic checks for number of servers bound)

**Comment by bajtos** [comment]
Thank you for an exhaustive explanation, @kjdelisle! ❤️ 

Here is my understanding:

`RestServer` does set `application` as the parent context AFAICT, see ][rest-server.ts](https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/rest-server.ts#L129-L133), all is good so far:

```ts
export class RestServer extends Context implements Server {
  // ...
  constructor(
    @inject(CoreBindings.APPLICATION_INSTANCE) app: Application,
    @inject(RestBindings.CONFIG) options?: RestServerConfig,
  ) {
    super(app);
    // ...
  }
}
```

`RestApplication#sequence` binds the sequence at application's level, see https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/rest-application.ts#L51-L53. Because each RestServer inherits all application's bindings, this should work . Unless RestServer has it's own sequence binding, in which case this server-level binding always overrides any application-level binding. Is this the case?

```ts
  sequence(sequence: Constructor<SequenceHandler>): Binding {
    return this.bind(RestBindings.SEQUENCE).toClass(sequence);
  }
```

`RestApplication#handler` is using a different approach. Instead of binding the handler at app-level, it retrieves the server and binds the handler there. See https://github.com/strongloop/loopback-next/blob/c86b0cf717966ccc4f58e305bc0179198576362c/packages/rest/src/rest-application.ts#L55-L59.

```ts
  handler(handlerFn: SequenceFunction) {
    // FIXME(kjdelisle): I attempted to mimic the pattern found in RestServer
    // with no success, so until I've got a better way, this is functional.
    const server = this.getSync('servers.RestServer');
    server.handler(handlerFn);
  }
```

I think this is pretty much what I was proposing for `app.route`. AFAICT, `app.route` API is not implemented by `RestApplication`, I think that's an oversight we should fix. Thoughts?

**Comment by bajtos** [comment]
> AFAICT, `app.route` API is not implemented by `RestApplication`, I think that's an oversight we should fix.

... we should fix as part of work on this story.

**Comment by kjdelisle** [comment]
If it's not a lot of effort, then that's fine. I don't want to hold this story up over what essentially amounts to a few lines of code, though I know that @jannyHou has some reservations about the direct definition of routes in the constructor, and I do feel a bit uneasy about that approach as well.

**Comment by jannyHou** [comment]
Discussed with @kjdelisle before: My concern for `app.route` is the top down approach is under the risk of inconsistency between controller and the spec:

Take the [jsdoc example](https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/rest-server.ts#L342-L347):

```js
class MyController {
  greet(name: string) {
    return `hello ${name}`;
  }
}
app.route('get', '/greet', operationSpec, MyController, 'greet');
```

If the `operationSpec.parameters` doesn't describe function `greet`'s arguments consistently, like:

```js
// operationSpec

{
  parameters: [
    {
      name: 'name',
      schema: {
         type: 'number' // But the name argument's type is string!
      }
    }
  ]
}
```
Then we get trouble when do the type coerce.

**Comment by bajtos** [comment]
> The top down approach is under the risk of inconsistency between the controller class and the spec:
> If the operationSpec.parameters doesn't describe function greet's arguments consistently
> Then we get trouble when do the type coerce.

IMO, this is a problem we need to solve as part of hardening the top-down approach and it's out of scope of this task, which is asking to simplify access to `route()` registration back to the version we used to have before support for multiple rest server complicated everything.

I also think the problem described above applies to bottom-up approach to, although it may be easier to spot.

```ts
class MyController {
  @get('/greet')
  @param.number('name');
  greet(name: string) {
    return `hello ${name}`;
  }
}
```

Or may be not, because both top-down and bottom-up examples above have another problem - the controller says `name` is required, but param spec does not set `{required: true}`, therefore the parameter is considered as optional in the API spec.

This mismatch between required/optional parameters in TS source and Swagger/OpenAPI decorators is a problem of many (if not all of) our current examples. However, our REST layer does not implement "required" validation yet (IIRC), therefore I think it does not matter much at this point.

**Comment by bajtos** [comment]
I went ahead and implemented `app.route()` and `app.api()`, please review: https://github.com/strongloop/loopback-next/pull/994

It took me less than 30 minutes to make that PR, so at least the initial effort was small 😄 Let's see how the review goes 🤞

**Comment by bajtos** [comment]
@shimks once #994 is landed, we need to edit documentation on loopback.io.

For example, http://loopback.io/doc/en/lb4/Application.html

```ts
export class WidgetApplication extends Application {
  constructor() {
    // This is where you would pass configuration to the base constructor
    // (as well as handle your own!)
    super();
    const app = this; // For clarity.
    // You can bind to the Application-level context here.
    // app.bind('foo').to(bar);
    app.component(RestComponent);
    app.controller(SamoflangeController);
    app.controller(DoohickeyController);
  }

  async start() {
    // This is where you would asynchronously retrieve servers, providers and
    // other components to configure them before launch.
    const server = await app.getServer(RestServer);
    server.bind('rest.port').to(8080);
    server.api(WidgetApi);
    // The superclass start method will call start on all servers that are
    // bound to the application.
    return await super.start();
  }
```

1. `WidgetApi` should be defined in the constructor via `this.api`.
2. Ideally, `rest.port` should be defined in the constructor too. We cannot do `this.bind('rest.port').to(8080)` because right now, `RestServer` overrides app-level host/port bindings it its constructor - see https://github.com/strongloop/loopback-next/blob/baf9928613ccde82c02abbc8249f4163e6a5570b/packages/rest/src/rest-server.ts#L142-L143
I think we have to modify `WidgetApplication`'s constructor to set the port via `ApplicationOptions#rest`.

That's another thing to improve in the way how RestServer interacts with RestApplication, I am not sure if we have bandwidth to address it as part of this story though.

Please grep the documentation for any other places where we configure `RestServer` from application's `start()` method, those places need to be fixed too.
